rskj-core/src/main/java/org/ethereum/core/BlockHeader.java

```diff
@@ -539,28 +533,29 @@ public String getParentPrintableHash() {
     public byte[] getMiningForkDetectionData() {
         if(includeForkDetectionData) {
             if (hasMiningFields() && miningForkDetectionData.length == 0) {
-                byte[] encodedBlock = getEncoded(false, false);
-                byte[] hashForMergedMining = HashUtil.keccak256(encodedBlock);
-
-                byte[] hashForMergedMiningPrefix = Arrays.copyOfRange(
-                        hashForMergedMining,
-                        0,
-                        HASH_FOR_MERGED_MINING_PREFIX_LENGTH
-                );
+                byte[] hashForMergedMining = getBaseHashForMergedMining();
+
                 byte[] coinbaseTransaction = getBitcoinMergedMiningCoinbaseTransaction();
 
-                List<Byte> hashForMergedMiningPrefixAsList = Arrays.asList(ArrayUtils.toObject(hashForMergedMiningPrefix));
-                List<Byte> coinbaseAsList = Arrays.asList(ArrayUtils.toObject(coinbaseTransaction));
+                byte[] mergeMiningTagPrefix = Arrays.copyOf(RskMiningConstants.RSK_TAG, RskMiningConstants.RSK_TAG.length + HASH_FOR_MERGED_MINING_PREFIX_LENGTH);
+                arraycopy(hashForMergedMining, 0, mergeMiningTagPrefix, RskMiningConstants.RSK_TAG.length, HASH_FOR_MERGED_MINING_PREFIX_LENGTH);
 
-                int position = Collections.lastIndexOfSubList(coinbaseAsList, hashForMergedMiningPrefixAsList);
+                int position = ListArrayUtil.lastIndexOfSubList(coinbaseTransaction, mergeMiningTagPrefix);
                 if (position == -1) {
                     throw new IllegalStateException(
                             String.format("Mining fork detection data could not be found. Header: %s", getPrintableHash())
                     );
                 }
 
-                int from = position + HASH_FOR_MERGED_MINING_PREFIX_LENGTH;
+                int from = position + RskMiningConstants.RSK_TAG.length + HASH_FOR_MERGED_MINING_PREFIX_LENGTH;
                 int to = from + FORK_DETECTION_DATA_LENGTH;
+
+                if (coinbaseTransaction.length < to) {
+                    throw new IllegalStateException(
+                            String.format("Invalid fork detection data length. Expected: %d. Got: %d. Header: %s", FORK_DETECTION_DATA_LENGTH, FORK_DETECTION_DATA_LENGTH - (to - coinbaseTransaction.length), getPrintableHash())
```

Comment:
Shouldn't `FORK_DETECTION_DATA_LENGTH - (to - coinbaseTransaction.length)` (the `Got: %d` bit) be `coinbaseTransaction.length - from - 1`?

Reply:
Results are the same. However, the other way (`length - from`) is way clearer, so I'll change it to that.

Reply:
Thanks